packages/graphql-explorer-html/src/App.tsx

```diff
@@ -1,5 +1,101 @@
-import React from 'react';
+import ApolloClient from 'apollo-boost';
+import config from 'graphql-explorer/lib/config';
+import { addCommonFields } from 'graphql-explorer/lib/inputFields';
+import addRelayTypes from 'graphql-explorer/lib/relay';
+import { deserializeQueryBuilder } from 'graphql-explorer/lib/serialization';
+import FieldPanel from 'graphql-explorer/lib/ui/FieldPanel';
+import Panels from 'graphql-explorer/lib/ui/Panels';
+import RootQueryPanel from 'graphql-explorer/lib/ui/RootQueryPanel';
+import qs from 'query-string';
+import React, { ComponentProps, useEffect, useMemo, useState } from 'react';
+import Navbar from 'react-bootstrap/Navbar';
 
-export default function App() {
-  return <div>empty</div>;
+import ConnectionParamsPanel, {
+  ConnectionParams,
+} from './ConnectionParamsPanel';
+import useSchema from './useSchema';
+
+type Location = 'ROOT_QUERY' | { query: string };
+
+type PanelProps = ComponentProps<typeof RootQueryPanel>;
+
+function App() {
+  const [connectionParams, setConnectionParams] = useState<ConnectionParams>({
+    uri: 'https://api.graph.cool/simple/v1/swapi',
+  });
+
+  const client = useMemo(
+    () => (connectionParams ? new ApolloClient(connectionParams) : null),
+    [connectionParams],
+  );
+
+  const schemaState = useSchema(connectionParams);
+  const [configured, setConfigured] = useState(false);
+  useEffect(() => {
+    if (schemaState.status === 'resolved') {
+      const { schema } = schemaState;
+      config.setSchema(schema);
+      addCommonFields();
+      addRelayTypes();
+
+      setConfigured(true);
+    }
+  }, [schemaState]);
+
+  const location: Location = useMemo(() => {
+    const queryString = qs.parse(window.location.search);
+    if (typeof queryString.query === 'string')
+      return { query: queryString.query };
+
+    return 'ROOT_QUERY';
+  }, []);
+
+  const rootPanel = useMemo(() => {
```

Comment:
are you able to type `useMemo` instead of the functions it returns? that'd be nifty

Reply:
uh? this code is fully typed